Fix clang build

hcorion opened the pull request:
This fixes the clang build and makes the travis tests go green!
Ignore the weird commit history.

hcorion committed:
Initial CMake fix

hcorion committed:
Fix compilation with GCC

hcorion committed:
Fix merge conflict

hcorion committed:
Get rid of awful hack

hcorion committed:
Update cotire with qt support

hcorion committed:
Maybe fix travis

hcorion committed:
Emergency Hack Relief Program Activated

hcorion committed:
Merge branch 'RPCS3-QT' of https://github.com/flash-fire/rpcs3 into fix-qt

hcorion committed:
Merge branch 'RPCS3-QT' of https://github.com/flash-fire/rpcs3 into fix-qt

hcorion committed:
fix clang build

hcorion committed:
fix clang build: The Empire Strikes Back

hcorion committed:
fix clang build: Return of the Jedi

hcorion committed:
Fix tab

Megamouse commented:
neko complained about new functions being introduced. thats why I made lambdas out of everything.
you could also simply use:

inline auto Pause = {blabla}

if that works with clang and travis

hcorion commented:
I'll try that @Megamouse

hcorion committed:
Merge branch 'RPCS3-QT' of https://github.com/flash-fire/rpcs3 into fix-qt

hcorion commented:
@Megamouse I'm having a hard time putting your words into code. Do you mean code like this?
```c++
inline auto Pause = 
{
	if (Emu.IsReady()) Emu.Run();
	else if (Emu.IsPaused()) Emu.Resume();
	else if (Emu.IsRunning()) Emu.Pause();
	else if (!Emu.GetPath().empty()) Emu.Load();
};
```
because that is not working, I haven't worked with lambdas before, so this is very new to me.

hcorion committed:
Fix random compilation issue

hcorion committed:
Final fix for clang, I think I got the lambda to work

hcorion committed:
Pretty diff is pretty

hcorion commented:
nvm, I think I got it working with lambdas.
@Megamouse Do you want to make sure my changes are doing what they are supposed to do?

Megamouse commented:
sorry I had a typo ;( . the right thing would look like this (hopefully without typo this time):

inline auto Pause() = 
{
	if (Emu.IsReady()) Emu.Run();
	else if (Emu.IsPaused()) Emu.Resume();
	else if (Emu.IsRunning()) Emu.Pause();
	else if (!Emu.GetPath().empty()) Emu.Load();
}

adn you would use it exactly like the lambda: 
connect(someButton, &QAbstractButton::clicked, Pause);

but I guess if lambda works now, it's good. help me understand: why the "this" in connect? that should normally work without it. And why was the [=] a problem? I'm gonna try this later, gotta go to brunch now

hcorion commented:
**Disclaimer:** I don't know C++/Qt very well.


> And why was the [=] a problem?

The clang build failed with the following error: 
```
rpcs3/rpcs3qt/main_window.cpp:82:15: error: non-local lambda expression cannot have a capture-default
auto Pause = [=]() {
              ^
```
The strange thing is that neither GCC or MSVC cared.


> help me understand: why the "this" in connect?

TBH, I'm not sure, I think it won't compile without `this` in the connect, but I didn't re-test. It was recommended by @flash-fire when I was converting it to a function and I just left it in when I converted it back to a lambda.

> sorry I had a typo ;( . the right thing would look like this (hopefully without typo this time):
>
> inline auto Pause() =

I tried that code as well but couldn't get it to build. The new code I just added seemed to work for some strange reason though.

Megamouse commented:
the "this" is definitely not needed in that case. you only need it if you want a specific object to execute one of its functions / slots.

hcorion committed:
get rid of 'this'

hcorion committed:
Merge branch 'RPCS3-QT' of https://github.com/flash-fire/rpcs3 into fix-qt

hcorion committed:
Fix random build breakage

hcorion committed:
fix breakage thanks to MegaMouse

hcorion committed:
Them tabs man
